src/Compilers/CSharp/Test/Semantic/Semantics/NullCheckedParameterTests.cs

```diff
@@ -170,5 +170,295 @@ class C
                     //     public static extern int M(int x!);
                     Diagnostic(ErrorCode.ERR_MustNullCheckInImplementation, "x").WithArguments("x").WithLocation(6, 36));
         }
+
+        [Fact]
+        public void NullCheckedMethodDeclaration()
+        {
+            var source = @"
+class C
+{
+    void M(string name!) { }
+}";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void FailingNullCheckedArgList()
+        {
+            var source = @"
+class C
+{
+    void M()
+    {
+        M2(__arglist(1, 'M'));
+    }
+    void M2(__arglist!)
+    {
+    }
+}";
+            CreateCompilation(source).VerifyDiagnostics( 
+                    // (8,22): error CS1003: Syntax error, ',' expected
+                    //     void M2(__arglist!)
+                    Diagnostic(ErrorCode.ERR_SyntaxError, "!").WithArguments(",", "!").WithLocation(8, 22));
+        }
+
+        [Fact]
+        public void NullCheckedArgList()
+        {
+            var source = @"
+class C
+{
+    void M()
+    {
+        M2(__arglist(1!, 'M'));
+    }
+    void M2(__arglist)
+    {
+    }
+}";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCheckedMethodValidationWithOptionalParameter()
+        {
+            var source = @"
+class C
+{
+    void M(string name! = null) { }
```

Comment:
I personally think this should be a build error. This signature says "you don't need to provide a value for this parameter, but we'll throw if you don't provide a value for it." There's no OHI concern here as with NRT.

It looks like [the spec](https://github.com/dotnet/csharplang/blob/master/proposals/null-arg-checking.md) doesn't address this.

Reply:
Please add a test or TODO for a non-null default parameter value. (I think literal strings are the only "good" case that could be used, actually..)

Reply:
Yeah this does look like a spec hole. Seems reasonable to make it an error, although I agree that it's out of scope for this PR and we can file an issue or make a PROTOTYPE comment and get back to it.

Reply:
Will add prototype comment and string literal opt param test.